octobot_trading/personal_data/portfolios/value_converter.py

```diff
@@ -0,0 +1,360 @@
+                    return bridge_part_1_value * bridge_part_2_value
+            except errors.MissingPriceDataError:
+                # conversion pairs might not be initialized or is not available at all
+                self._ensure_no_pending_symbol_price(part_1_quote, part_2_quote)
+                # otherwise continue with other pairs
+            bridge = base_bridge + [(part_1_base, part_1_quote)]
+            # case 2: X parts bridge
+            nested_value = self.try_convert_currency_value_using_multiple_pairs(
+                part_1_quote, target, constants.ONE, bridge
+            )
+            if nested_value:
+                try:
+                    extended_bridge = [(part_1_base, part_1_quote)] \
+                        + self.get_saved_price_conversion_bridge(part_1_quote, target)
+                    self._save_price_bridge(currency, target, extended_bridge)
+                except KeyError:
+                    # should not happen, however if it does, do not crash
+                    pass
+                return bridge_part_1_value * nested_value
+        # no bridge found
+        self._save_missing_price_bridge(currency, target)
+        return None
+
+    def _get_priced_pairs(self):
+        for pair in self.last_prices_by_trading_pair:
+            # first look into pairs with price
+            yield pair
+        for pair in self.portfolio_manager.exchange_manager.exchange_config.traded_symbol_pairs:
+            # then into pairs from config
+            if pair not in self.last_prices_by_trading_pair:
+                yield pair
+        if self.initializing_symbol_prices:
+            for pair in self.initializing_symbol_prices_pairs:
+                # finally into initializing pairs
+                yield pair
+
+    def _ensure_no_pending_symbol_price(self, base, quote):
+        for symbol in (symbol_util.merge_currencies(base, quote), symbol_util.merge_currencies(quote, base)):
+            if symbol in self.portfolio_manager.exchange_manager.exchange_config.traded_symbol_pairs \
+                    or symbol in self.initializing_symbol_prices_pairs:
+                raise errors.PendingPriceDataError
+
+    def get_saved_price_conversion_bridge(self, currency, target) -> list:
+        return self._price_bridge_by_symbol[symbol_util.merge_currencies(currency, target)]
+
+    def _save_price_bridge(self, currency, target, bridge):
+        self._price_bridge_by_symbol[symbol_util.merge_currencies(currency, target)] = bridge
+
+    def convert_currency_value_from_saved_price_bridges(self, currency, target, quantity) -> decimal.Decimal:
+        try:
+            bridge = self._price_bridge_by_symbol[symbol_util.merge_currencies(currency, target)]
+            converted_value = quantity
+            for base, quote in bridge:
+                converted_value = self.convert_currency_value_using_last_prices(converted_value, base, quote)
```

Comment:
I don't understand this part, why do we ignore all converted_value except the last one?

Reply:
all previous ones are saved into converted_value  that is reused each time